### DISCUSSION TITLE
아이템 9. try-finally보다는 try-with-resources를 사용하라

### DISCUSSION THREAD

**Discussion by coalong** [Study-2-Effective-Java/Effective-Java#28] [3. 과제] [answered]
## 아이템 9. try-finally보다는 try-with-resources를 사용하라

- 자바 라이브러리에는 close 메서드를 호출해 직접 닫아줘야 하는 자원이 많다.
ex) InputStream, OutputStream, java.sql.Connection 등
- Java Application 내에서만 사용되는 자원들이 아니라 운영 체제의 자원을 사용하는 것들
ex) File, Network, DB 등

### 3. **try-finally (Java 7 이전의 finally 방식)**  ⭐ 
<details>
<summary>예시 코드</summary>
<div markdown="1">

- MyResource 클래스
```java
        public class MyResource implements AutoCloseable{
        
            public void doSNothing(){
                System.out.println("Nothing");
            }
        
            public void doSomething(){
                System.out.println("Do Something");
                throw new FirstError();
            }
        
            public void doSomething1(){
                System.out.println("Do Something1111");
                throw new FirstError();
            }
        
            @Override
            public void close() throws SecondError {
                System.out.println("Close My Resource");
                throw new SecondError();
            }
        }
```
-  FirstError 클래스
```java
        public class FirstError extends RuntimeException{
            public FirstError() {
                super("첫 번째 에러입니다.");
            }
        }
```
- SecondError 클래스
```java
        public class SecondError extends RuntimeException {
            public SecondError() {
                super("두 번째 에러입니다.");
            }
        }
```
</div>
</details>

- try-finally 사용 예시
```java
    public class AppRunner {
        public static void main(String[] args) {
            MyResource myResource = new MyResource();
            try {
                myResource.doSomething();
            } finally {
                myResource.close();
            }
        }
    }
```
 
<details>
<summary>실행결과</summary>
<div markdown="1">

```
Do Something
Close My Resource
Exception in thread "main" org.example.ex1.SecondError: 두 번째 에러입니다.
	at org.example.ex1.MyResource.close(MyResource.java:13)
	at org.example.ex1.AppRunner.main(AppRunner.java:9)

Process finished with exit code 1
```

</div>
</details>

`문제점 1.` 

- 예외는 try 블록과 finally 블록 모두에서 발생할 수 있는데, 만약 이런 상황이라면 두 번째 예외가 첫 번째 예외를 완전히 집어삼켜 버린다. (흔적도 없이.)
- 스택 추적 내역에 첫 번째 예외에 관한 정보는 남지 않게 되어, 실제 시스템에서의 디버깅이 몹시 어렵다.
- 물론 첫 번째 예외를 기록하도록 코드를 수정할 수는 있지만, 코드가 너무 지저분해져서 실제로 그렇게 까지 하는 경우는 거의 없다.
    
`문제점 2.`

- close 해야 할 여러 자원을 사용하게 되면 자원 해제하지 못하는 위험 가능성이 있다.
    
    ```java
    public class AppRunner {
        public static void main(String[] args) {
            MyResource myResource = new MyResource();
            MyResource myResource1 = new MyResource();
            try {
                myResource.doSNothing();
                myResource1.doSomething1();
                }
             finally {
                myResource.close();   // 여기서 예외 발생하면 myResource1은 close 되지 않는다.
                myResource1.close();
            }
        }
    }
    ```
    
<details>
<summary>실행결과</summary>
<div markdown="1">

```
Nothing
Do Something1111
Close My Resource
Exception in thread "main" org.example.ex1.SecondError: 두 번째 에러입니다.
	at org.example.ex1.MyResource.close(MyResource.java:22)
	at org.example.ex1.AppRunner.main(AppRunner.java:12)

Process finished with exit code 1
```
    
[참고자료 stackoverflow](https://stackoverflow.com/questions/48449093/what-is-wrong-with-this-java-puzzlers-piece-of-code)

</div>
</details>

`문제점 3.`
- close() 하는 순서가 중요한 자원일 경우, 코드가 길어지고 순서를 잘 못 처리 했을 시 에러 발생 가능
- 아래 예시는 JDBC Connection 하는 일부 코드이다.

```java
public class Class1 {
    public method1() throws Exception {

       conn = DBPool.getConnection(); //
        stmt = conn.createStatement();
        rs = stmt.executeQuery(..);

      try {
          conn = ...;
          ps = ...;
          rs = ...;
      } catch (Exception e) {
          // ...
      } finally {
          if (rs != null) try { rs.close(); } catch(Exception e) {}
          if (ps != null) try { ps.close(); } catch(Exception e) {}
          if (conn != null) try { conn.close(); } catch(Exception e) {}
      }
    
    }
}
```
- 일반적으로 JDBC 객체를 얻는 순서는 Connection -> Statement -> ResultSet 순이며, 객체를 닫는 순서는 ResultSet -> Statement -> Connection 순이다. 즉, 먼저 얻은 객체를 가장 마지막에 닫는다. 
- 프로그래머의 실수로 이 순서에 맞게 close() 하지 않을 경우 예외가 발생할 수 있다.
    

### 4. **try-with-resources (Java 7)** ⭐ 

- try-with-resources 구조를 사용하려면 해당 자원이 AutoCloseable 인터페이스를 구현해야 한다. 

<details>
<summary> AutoCloseable ❓ </summary>
<div markdown="1">

```mermaid   

classDiagram
      
      AutoCloseable <|--  Closeable

      interface Closeable
         
      
      
```

- AutoCloseable 
```java
public interface AutoCloseable {
    void close() throws Exception;
        }
```
- Java 7 부터 도입
     - try-with-resources 문으로 관리되는 객체일 때 close() 메서드가 자동으로 호출된다. (JVM이 close() 를 호출)
     - close 메서드 구현시 구체적인 exception 을 throw 하고, close 동작이 전혀 실패할 리가 없을 때만 exception 을 던지지 않도록 구현하는 것을 강력히 권고한다.
     - 특히 close 메서드에서 InterruptedException을 던지지 않는 것을 권고한다.
     - InterruptedException은 쓰레드의 인터럽트 상태와 상호작용 하므로 interruptedException이 억제되었을 때 런타임에서 잘못된 동작이 발생할 수 있기 때문이다.
     - `멱등성(idempotent)`을 유지하는게 필수적인 사항은 아니지만 유지할 수 있도록 메서드를 구현하는 것을 추천한다.

- Closeable
```java
public interface Closeable extends AutoCloseable {
    public void close() throws IOException;
}
```
- Java 5 에 도입
     - Java 5 에서 I/O 관련 클래스들의 close() 를 하기 위해 인터페이스로 관리하여 사용
     - 하위 호환성 유지를 위해 남겨져 있고, I/O 관련 클래스들을 처리하기에 예외를 IOException 으로 던진다.
     - AutoCloseable 인터페이스를 상속받고 있어 Java 7 버전 이후로는 try-with-resources 문에 사용하면 자동으로 close() 메서드 호출된다.
     - 이 인터페이스는 `멱등성(idempotent)`을 유지하도록 메서드를 구현하는 것이 필수적이다. 따라서 몇 번을 호출하든 부작용(side effect)이 없다.

❗ 정리
- Closeable 과 AutoCloseable 은 try-with-resources 안에서 거의 동일한 기능을 한다.
- Closeable 을 구현할 때는 close 메소드의 멱등성이 필수조건이고, AutoCloseable 은 권장사항이다.
    - 이유는? 아마 Closeable 인터페이스를 사용할 당시 (Java 5) 에는 close() 를 수동으로 닫아줬기 때문에 여러번 호출할 위험 가능성이 있어서 라고 생각한다. 하지만 AutoCloseable 도 멱등성을 유지하도록 구현하는 것이 권장사항.
- 하위 호환성을 위해 Closeable 을 남겨둔 채로 AutoCloseable 을 상속받았기 때문에
   만약, close() 해야될 클래스를 구현할 경우 AutoCloseable 을 사용하는 것이 더 바람직하다.

</div>
</details>

- try-with-resources 사용 예시 1.
```java
    public class AppRunner {
        public static void main(String[] args) {
            try (MyResource myResource = new MyResource()) {
                myResource.doSomething();
            }
        }
    }
```
    
<details>
<summary>실행결과</summary>
<div markdown="1">

```
Do Something
Close My Resource
Exception in thread "main" org.example.ex1.FirstError: 첫 번째 에러입니다.
	at org.example.ex1.MyResource.doSomething(MyResource.java:11)
	at org.example.ex1.AppRunner.main(AppRunner.java:6)
	Suppressed: org.example.ex1.SecondError: 두 번째 에러입니다.
		at org.example.ex1.MyResource.close(MyResource.java:22)
		at org.example.ex1.AppRunner.main(AppRunner.java:5)

Process finished with exit code 1
```
  - doSomething 메서드 실행 도중 예외가 발생해도 close() 된다.    
  - try-finally에 비해 훨씬 간결해진 코드
  - 예외가 몇번 발생해도 삼켜지지 않고 모두 다 기록된다.
  - 실행결과에서 볼 수 있듯이 프로그래머에게 보여줄 예외는 하나만 보존되고, 여러 개의 다른 예외가 숨겨질 수 있다. 이렇게 숨겨진 예외들은 suppressed 라는 꼬리표를 달고 출력된다.

</div>
</details>
 
 
- try-with-resources 사용 예시 2.
```java
    public class AppRunner {
        public static void main(String[] args) {
            try (MyResource myResource = new MyResource();
                 MyResource myResource1 = new MyResource()) {
                myResource.doSomething();
                myResource1.doSomething();
            }
        }
    }
```

<details>
<summary>실행결과</summary>
<div markdown="1">

```
Do Something
Close My Resource
Close My Resource
Exception in thread "main" org.example.ex1.FirstError: 첫 번째 에러입니다.
	at org.example.ex1.MyResource.doSomething(MyResource.java:11)
	at org.example.ex1.AppRunner.main(AppRunner.java:7)
	Suppressed: org.example.ex1.SecondError: 두 번째 에러입니다.
		at org.example.ex1.MyResource.close(MyResource.java:22)
		at org.example.ex1.AppRunner.main(AppRunner.java:5)
	Suppressed: org.example.ex1.SecondError: 두 번째 에러입니다.
		at org.example.ex1.MyResource.close(MyResource.java:22)
		at org.example.ex1.AppRunner.main(AppRunner.java:5)

Process finished with exit code 1
```

- 처음 실행하는 doSomething() 에서 예외가 발생하더라도 사용한 모든 자원을 close()한다.
- 예외가 몇 번 발생하더라도 발생한 모든 기록이 남는다.

</div>
</details>
 

➕  또한, Java 7에서 추가된 Throwable 의 getSuppressed 메서드를 이용하면 프로그램 코드에서 가져올 수 있다.

```java
public class AppRunner {
    public static void main(String[] args) {
        try (MyResource myResource = new MyResource()) {
            myResource.doSomething();
        } catch (Throwable e) {
            Throwable[] suppExe = e.getSuppressed();
            System.out.println("Suppressed Exception Array length = " + suppExe.length);
            System.out.println("Message Text = " + suppExe[0].getMessage());
        }
    }
}
```

<details>
<summary>실행결과</summary>
<div markdown="1">

```
Do Something
Close My Resource
Suppressed Exception Array length = 1
Message Text = 두 번째 에러입니다.

Process finished with exit code 0
```
- Suppressed(억제된) Exception을 확인 할 수 있다.

</div>
</details>

    
### 5. **항샹된 try-with-resources (Java 9)** ⭐ 
    
```java
    public class AppRunner {
        public static void main(String[] args) {
            MyResource myResource = new MyResource();
            try (myResource) {
                myResource.doSomething();
            }
        }
    }
```
    
- try 블럭 외부에서 생성된 객체를 참조하여 auto close 가 가능하여 try 블럭의 가독성을 높일 수 있다.

## 중첩된 자원의 close() 는 어떻게 처리 될까?
- 예시 코드
```java
try (BufferedReader reader = new BufferedReader(new InputStreamReader(socket.getInputStream()))) {
		
} catch (IOException e) {

}
```
- try() 안에 중첩적인 스트림 객체를 생성 한다면?
- BufferedReader 외의 자원들은 전부 자동으로 닫힐까?
<details>
<summary>해답</summary>
<div markdown="1">
        

:white_small_square: 해답은 try() 문에서 자신의 변수가 아닌 객체의 close()를 자동으로 해주진 않는다.   
:white_small_square:  Stream 클래스들이 객체 내에서 자신의 필드에 대한 close()도 호출해주기 때문이다.
:white_small_square: 만약 AutoCloseable 의 구현체를 생성할때는 해당부분도 유의해서 작성하자.

</div>
</details>

---
<aside>
💡 핵심정리   

- 꼭 회수해야 하는 자원을 다룰 때는 try-finally 말고, try-with-resources를 사용하자.   
- 예외는 없다. 코드는 더 짧고 분명해지고, 만들어지는 예외 정보도 훨씬 유용하다.

</aside>

---
### 참고자료
- [Docs - Class BufferedReader](https://docs.oracle.com/javase/8/docs/api/java/io/BufferedReader.html#close--)
- [Docs - Class InputStreamReader](https://docs.oracle.com/javase/8/docs/api/java/io/InputStreamReader.html#close--)
- [참고자료](https://multifrontgarden.tistory.com/192)
- [참고자료](https://stackoverflow.com/questions/1388602/do-i-need-to-close-both-filereader-and-bufferedreader)
- [참고자료](https://www.baeldung.com/java-try-with-resources)
- [참고자료](https://androphil.tistory.com/763)

**Comment by JoisFe** [comment, marked as answer]
정리 감사합니다. try-with-resource 구조를 사용하려면 해당 자원이 AutoCloseable 인터페이스를 구현해야 한다. 라고 하셨는데 Closeable 인터페이스를 구현한 객체 또한 try-with-resource 구조를 사용할 수 있는 것으로 알 고 있습니다.
AutoCloseable과 Closeable의 차이가 무엇일까요?

![image](https://user-images.githubusercontent.com/90208100/209426955-72c3b568-5568-4ea0-9b54-86f444ca334b.png)

**Reply by chikeem90** [reply]
오,, 저도 궁금해서 좀 서칭을 해봤는데,,
Closeable은 java 7 이전 I/O 관련 클래스들의 인터페이스로 사용되었던 거 같습니다.
7버전 이전에는 명시적으로 close를 호출했어야 했는데, 그 때 Closeable을 구현한 close 메서드를 호출했던 것이죠.
https://docs.oracle.com/javase/6/docs/api/
근데 이제 java 7에 TWR (try-with-resources) 가 추가되면서 AutoCloseable 인터페이스를 추가했고, 기존의 코드와 하위호환성을 위해 Closeable도 그대로 유지한 채 AutoCloseable을 상속하는 형태가 된 거 같습니다.

```java
public interface AutoCloseable {
    void close() throws Exception;
}
```

```java
public interface Closeable extends AutoCloseable {
    public void close() throws IOException;
}
```

그래서 거의 동일한 인터페이스라고 보면 될 거 같은데,, 미묘한 차이가 있긴 하네요.
앞서 언급했듯 I/O 관련 클래스들의 인터페이스였기 때문에 throw하는 Exception이 IOException으로 한정되어 있습니다.
그래서 사실 커스텀하게 구현해서 쓰기에는 AutoCloseable이 더 유용할 거 같구요.
또한, 자바 명세에 정의되어 있는게, Closeable의 close는 멱등성이 필수조건이고, AutoCloseable은 권장?정도 이네요.
https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/AutoCloseable.html

**Reply by coalong** [reply]
@JoisFe 놓치고 있던 부분을 알려주셔서 감사합니다! 
@chikeem90 님이 댓글로 자세한 내용과 자료링크를  남겨주셔서 바탕으로 찾아보고 
다시 위에 내용을 정리했습니다.!!
👍 👍 👍

**Reply by JoisFe** [reply]
@chikeem90 답변 감사합니다.
@coalong 다시 정리해주신 부분도 잘 읽어보겠습니다 감사합니다.

**Comment by Irisation23** [comment]
정리 해 주신글 잘 읽었습니다 👍
> `항샹된 try-with-resource (Java9)` 의 장점이 가독성이 높다라는 설명을 해주셨습니다.
> - `try-with-resource (Java7)` 보다 좋은 점이 무엇이 있을까요?? 🤔

![image](https://user-images.githubusercontent.com/53285909/209427190-d5cbb8c6-e7f4-4f83-9115-beab3313a6cc.png)

**Reply by chikeem90** [reply]
> try-with-resource (Java7) 보다 좋은 점이 무엇이 있을까요?? 🤔

https://www.javatpoint.com/java-9-try-with-resources
https://docs.oracle.com/en/java/javase/19/language/java-language-changes.html#GUID-A920DB06-0FD1-4F9C-8A9A-15FC979D5DA3

try 블럭 외부에서 선언한 변수의 경우 try 블럭에서 바로 접근하지 못하고 try 문에서 로컬 변수로 한번 더 선언해서 써야했으나, final이거나 effective final인 경우에는 그냥 바로 try-with-resources 로 가능하네요.

```java
    public void testOne() throws FileNotFoundException {
        BufferedReader brs = new BufferedReader(new FileReader("C:\\number.txt"));
        try (BufferedReader br = brs) { // try 블럭 로컬 변수에 할당
            String str;
            while ((str = br.readLine()) != null) {
                System.out.println(str);
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
```

```java
    public void testOne() throws FileNotFoundException {
        BufferedReader br = new BufferedReader(new FileReader("C:\\number.txt"));
        try (br) { // 바로 사용 가능
            String str;
            while ((str = br.readLine()) != null) {
                System.out.println(str);
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
```

장점은,, 좀 더 간결해졌다,,? ㅎㅎ 불필요하게 로컬 변수를 선언할 필요 없다,,?

**Reply by coalong** [reply]
네! 저도 장점이라고 하면 try() 안에 길어지는 코드나 auto close 할 객체가 여러 개 일 경우 가독성 측면에서 좋을 수 있겠다 생각했어요!

```java
public class AppRunner {
    public static void main(String[] args) {

        MyResource myResource1 = new MyResource();
        MyResource myResource2 = new MyResource();
        MyResource myResource3 = new MyResource();

        try (myResource1; myResource2; myResource3) {
            myResource1.doSomething();
            myResource2.doSomething();
            myResource3.doSomething();
        }
    }
}
```

@chikeem90 님이 위에서 설명과 자료를 자세히 첨부해주셔서 감사합니다!

**Reply by Irisation23** [reply]
`읽기 좋은 코드가 좋은 코드` 라는 책이 생각나네요.
정리 감사합니다 @chikeem90 님 @coalong 님 👍

**Comment by corock** [comment]
잘 읽었습니다~! 몇 가지 본 내용을 공유드리자면,

- 아마 책에서는  _try-with-resource**s**_ 처럼 복수형으로 용어를 소개하고 있어서 교정된다면 좋을 것 같아요.
- 저는 자원을 `close()` 순서가 어긋나 오류가 난 경험이 있어 *try-with-resources* 적극 사용하는 편입니다.

**Reply by coalong** [reply]
감사합니다.!
해당 내용 수정했습니다!
